strawberry/schema/schema.py

```diff
@@ -214,15 +214,64 @@ class Query:
             formatted_errors = "\n\n".join(f"❌ {error.message}" for error in errors)
             raise ValueError(f"Invalid Schema. Errors:\n\n{formatted_errors}")
 
-    def get_extensions(
-        self, sync: bool = False
-    ) -> List[Union[Type[SchemaExtension], SchemaExtension]]:
-        extensions = list(self.extensions)
+    def get_extensions(self, sync: bool = False) -> List[SchemaExtension]:
+        if not self._cached_extensions:
+            extensions = []
+            if self.directives:
+                extensions = [
+                    *self.extensions,
+                    DirectivesExtensionSync if sync else DirectivesExtension,
+                ]
+            extensions.extend(self.extensions)
+            self._cached_extensions = [
+                ext if isinstance(ext, SchemaExtension) else ext(execution_context=None)
+                for ext in extensions
+            ]
+        return self._cached_extensions
+
+    def create_extensions_runner(
+        self, execution_context: ExecutionContext, extensions: list[SchemaExtension]
+    ) -> SchemaExtensionsRunner:
+        return SchemaExtensionsRunner(
+            execution_context=execution_context,
+            extensions=extensions,
+        )
+
+    def _get_middleware_manager(
+        self, extensions: list[SchemaExtension]
+    ) -> MiddlewareManager:
+        # create a middleware manager with all the extensions that implement resolve
+        if not self._cached_middleware_manager:
+            self._cached_middleware_manager = MiddlewareManager(
+                *(ext for ext in extensions if self._implements_resolve(ext))
+            )
+        return self._cached_middleware_manager
 
-        if self.directives:
-            extensions.append(DirectivesExtensionSync if sync else DirectivesExtension)
+    def _create_execution_context(
+        self,
+        query: Optional[str],
+        allowed_operation_types: Iterable[OperationType],
+        variable_values: Optional[Dict[str, Any]] = None,
+        context_value: Optional[Any] = None,
+        root_value: Optional[Any] = None,
+        operation_name: Optional[str] = None,
+    ) -> ExecutionContext:
+        return ExecutionContext(
+            query=query,
+            schema=self,
+            allowed_operations=allowed_operation_types,
+            context=context_value,
+            root_value=root_value,
+            variables=variable_values,
+            provided_operation_name=operation_name,
+        )
 
-        return extensions
+    @classmethod
+    def _implements_resolve(cls, obj: SchemaExtension | type[SchemaExtension]) -> bool:
+        """Whether the extension implements the resolve method."""
+        if isinstance(obj, SchemaExtension):
+            return type(obj).resolve is not SchemaExtension.resolve
+        return obj.resolve is not SchemaExtension.resolve
```

Comment:
nit: This could be a `SchemaExtension` classmethod

also, `return obj.resolve is not SchemaExtension.resolve` would fail with `AttributeError` if the obj doesn't have a resolve key in its dict